clippy_lints/src/use_self.rs

```diff
@@ -176,97 +268,21 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UseSelf {
                     true
                 };
 
+                // TODO: don't short-circuit upon lifetime parameters
                 if should_check {
-                    let visitor = &mut UseSelfVisitor {
-                        item_path,
+                    let self_ty= hir_ty_to_ty(cx.tcx, item_type);
```

Comment:
My guess is, that this is causing the ICE in `could_be_const.rs`. It's probably some wrong `hir_id` being passed in, but I'm not sure beyond that

Reply:
Indeed, commenting sections making use of this seems to solve failing test cases. 

Could you help me out on this? 

I've been trying to debug this without much success. The backtraces are not very helpful and the errors are raised from rustc location I have no understanding of. What is the `hir_ty_to_ty` contract? Under which circumstances could this call lead to errors elsewhere?

Reply:
Sure, I can try =) Beyond that, I'm not really sure about how to use `hir_ty_to_ty` properly and it's not documented either.
I can see two ways to move a bit forward though:

1. Could you try not doing the conversion for `def::Res::SelfTy`? That would look like this: https://github.com/rust-lang/rust-clippy/blob/f2486b3d35efc551ea0b5b549120c05f3ab1d636/clippy_lints/src/use_self.rs#L94-L98
2. `dbg!` as much as possible around this location and figure out if there's something that looks odd

Reply:
I think the problem is located in the `visit_ty` method of the visitor. See line 167 mentioned in another comment hereunder.